cpp/marc_tools/single_marc_convert_keywords.cc

```diff
@@ -59,69 +59,65 @@ void ExtractSubfieldsForTag(const MARC::Record &record, const std::string &subfi
-        return;
-    for (const char subfield_code : subfield_codes) {
-        std::string subfield_value(record.getFirstSubfieldValue(subfield_tag, subfield_code));
-        if (not subfield_value.empty() ) {
-            if (subfield_code == 'x')
-                subfield_value = '(' + subfield_value + ')';
-            subfields->emplace_back(subfield_value);
-        }
-        else if (subfield_code == 'a')
-            LOG_WARNING("Entry has no Subfield 'a' for PPN " + record.getControlNumber());
-    } */
+}
+
+
+void AddMainSubfieldAndCombinationsToGndKeywords(const MARC::Record &record, std::unordered_map<std::string, std::string> * const keywords_to_gnd_numbers_map, const std::string &field_tag, const std::string &subfield_codes, std::vector<std::string> * const subfields) {
+    subfields->clear();
+    // adding extra entry for every 'a' (main) subfield of every tag for a higher matching chance
+    if (subfield_codes.find('a') != std::string::npos) {
+        std::string subfield_value_a(record.getFirstSubfieldValue(field_tag, 'a'));
+        if (not subfield_value_a.empty())
+            keywords_to_gnd_numbers_map->emplace(std::make_pair(subfield_value_a, record.getControlNumber()));
+    }
+    ExtractSubfieldsForTag(record, field_tag, subfield_codes, subfields);
+    if (subfields->size() > 1) {
+        const std::string key(StringUtil::Join(*subfields, " "));
+        keywords_to_gnd_numbers_map->emplace(std::make_pair(key , record.getControlNumber()));
+    }
+}
 
 
-void ReadInGndKeywords(MARC::Reader * const marc_reader, std::unordered_map<std::string, std::string> * const gnd_keywords)
+void ReadInGndKeywords(MARC::Reader * const marc_reader, std::unordered_map<std::string, std::string> * const keywords_to_gnd_numbers_map)
 {
     unsigned record_count(0);
 
-    while (MARC::Record record = marc_reader->read()) {
+    while (const MARC::Record record = marc_reader->read()) {
         ++record_count;
 
         std::vector<std::string> subfields;
-        ExtractSubfieldsForTag(record, "150", "agx", &subfields);
-        // making 2 entries: one for every single 'a' subfield and one for all subfields
-        if (subfields.size() == 1)
-            gnd_keywords->emplace(std::make_pair(subfields[0], record.getControlNumber()));
-        ExtractSubfieldsForTag(record, "100", "atg", &subfields);
-        ExtractSubfieldsForTag(record, "110", "agx", &subfields);
-        ExtractSubfieldsForTag(record, "111", "agx", &subfields);
-        ExtractSubfieldsForTag(record, "130", "agx", &subfields);
-        ExtractSubfieldsForTag(record, "151", "agx", &subfields);
-        // second entry for all subfields combined
-        if (subfields.size() > 1) {
-            const std::string key(StringUtil::Join(subfields, " "));
-            gnd_keywords->emplace(std::make_pair(key , record.getControlNumber()));
-        }
+        AddMainSubfieldAndCombinationsToGndKeywords(record, keywords_to_gnd_numbers_map, "150", "agx", &subfields);
```

Comment:
You're not using "subfields" in this function so please take it out here and declare it as a local variable in AddMainSubfieldAndCombinationsToGndKeywords.